bootstrap.md

```diff
@@ -0,0 +1,136 @@
+
+This is obviously impractical, so we choose to start the bootstrap sequence with nodes by connecting with an inital set of trusted peers. These can be nodes run by well-known entities (the Filecoin foundation, large exchanges, good actors in the community, etc) or manually specified by the node operator in the config file. We leave trustless (or fully decentralized) bootstrapping for future work.
+
+A strawman bootstrapping sequence would have nodes use these initial peers to sync up to the canonical chain. However, new connection requests might effecitvely DoS these peers, making the Filecoin network unavailable to new nodes. Therefore, Filecoin bootstrapping seeks to "load-balance" initial requests to ensure the network remains widely available.
+
+We connect to each of these initial peers using the `hello handshake` (see in [network protocols](./network-protocols.md#hello-handshake)), getting back
+
+```go
+type HelloMessage struct {
+HeaviestTipSet []Cid
+HeaviestTipSetHeight uint64
+GenesisHash Cid
+}
+```
+
+followed by a `GetPeers` request, getting back
+
+```go
+type PeersMessage struct {
+PeerSet []PeerID
+}
+```
+
+We will use these new peers to sync with the chain, using the chain heads from our initial set of trusted nodes as a means of verifying that these peers are mining the canonical chain. These initial peers do not return their latest heads, but rather their latest confirmed heads, thereby ensuring other honest peers should be aware of these as well.
+
+We assume no long-lived network partitions in Filecoin, meaning these confirmed heads will all be on the same chain (though potentially at different heights). A node should initially connect to 5 trusted nodes to begin bootstrapping.
+
+```go
+type Bootstrapper struct {
+    trustedPeers []PeerID
+    trustedHeads map[[]Cid]uint64
+    genesisHash Cid
+    
+    bootstrapPeers []PeerID
+    
+}
+
+func InitialConnect(config ConfigFile, b Bootstrapper, s Syncer) {
+    b.trustedPeers = configFile.bootstrapPeers
+    
+    for trustedPeer := range b.trustedPeers {
+        trustedChain := sayHello(trustedPeer)
+        otherPeers := getPeers(trustedPeer).PeerSet
+        
+        b.trustedHeads[trustedChain.HeaviestTipSet] = trustedChain.HeaviestTipSetHeight
+    	
+        if b.genesisHash == nil {
+	        b.genesisHash = trustedChain.GenesisHash           
+        } else if b.genesisHash != trustedChain.GenesisHash {
+            panic("One of your trusted peers is bad.")
+        }
+        
+        b.bootstrapPeers = append(b.bootstrapPeers, otherPeers)
+    }
+    
+    s.genesis = b.genesisHash
+}
+```
+
+
+
+## Peer Set Expansion
+
+Once connected to some portion of the network, a node will need to expand their peer set in order to get a good sample of the network and a live latest head. To do so, use the DHT to do random `FindPeer` requests. We do so until we have requested 5 new peers from each peer in the PeerSet.
```

Comment:
I'm thinking that instead of using "random FindPeer requests", we should have a special dht query method called bootstrap that uses the existing underlying query code, but takes as input some 'trusted set' of peers, and uses this algorithm internally to bootstrap, perhaps even having it accept some function to help implement the filecoin specific functionality.

cc @raulk @anacrolix @stebalien

Reply:
It seems to me that this bootstrapping should be implemented at the FIL level actually: it seems to be pretty specific for something like libp2p (eg even the notion of a trusted set of peers). So FindPeer would be a way to do non-content-based peer retrieval (i.e. `FindRandomPeer`) composed in Filecoin (or whatever other user of libp2p) to make up a bootstrap (eg).

Reply:
My point was actually that this doesnt seem filecoin specific at all, and is a good general method for connecting to a distributed network given an initial set of peers you know you trust, a way of determining if there are divergent sets, and a mechanism for getting more peers.

Reply:
fair enough. implemented here in case, but happy to help move this to libp2p if better.